### PR TITLE
EZP-25717: Fix all alias types cleaning

### PR DESCRIPTION
**Pull request by galileo**
Fixes: https://jira.ez.no/browse/EZP-25717

> Status: TBD

Some status information, for more please watch discussion in #1639.

>So after a long time of investigation, because there were a lot of incosistency with my test behavior (once it worked, and the other time it wasn't) i got an conclusion that the methods are missbehaving on top of how many times there where called.

>So the problem is how many times we hit the ->lookup('/path') method, why?

> 1. The first call loads an urlAlias, and stores it in cache#stash_default#ez_spi#urlalias#url#media.
> 2. The second call, loads from cache cache#stash_default#ez_spi#urlalias#url#media and loads urlPath from persistence mechanism and stores to the cache the alias with cache#stash_default#ez_spi#urlalias#0-62933a2951ef01f4eafd9bdf4d3cd2f0# ID!!
> 3. For the third call we have all information in the caching mechanism so we don't call the persistence mechanism so if there were some changes we don't know about them, and we don't have any cleanups for this cache element (url, and alias-id) so even if we don't have the location we will always get the historical alias.
>So for now this is why my test wasn't consistent. I will now search for solution but to be honest the easiest solution would be to cleanup the whole 'urlAlias' from cache mechanism to be sure about the consistency and not have to deal with all those dependencies.

>Of course I don't know how does this affect the speed and the cache itself so If you have any opinion about this I will be very glad to here about it.

### CODE REVIEW

**Comment by galileo** [comment]
So I prepared some improvements based on our talk I think this is now working good, but I still have feeling that this is not best solution but just less bad than previous.

This looks really awkward: https://github.com/ezsystems/ezpublish-kernel/pull/1646/files#diff-708405a2d74fe0e5ec8f1d7608271d03R280

ping @andrerom

File: eZ/Publish/Core/Persistence/Cache/UrlAliasHandler.php [hunk 1]
```diff
@@ -205,7 +204,8 @@ public function lookup($url)
             try {
                 $this->logger->logCall(__METHOD__, array('url' => $url));
                 $urlAlias = $this->persistenceHandler->urlAliasHandler()->lookup($url);
-                $cache->set($urlAlias->id);
+                $urlAliasId = $urlAlias->id;
```
Comment (andrerom) [inline eZ/Publish/Core/Persistence/Cache/UrlAliasHandler.php]: Why the temp variable?

Also: you can warmup the cache here, but make that change a separate commit please, as it is a different topic, if you want, rebase PR and make that the first commit _(remember to adapt tests if needed for the commit)_.

File: eZ/Publish/Core/Persistence/Cache/UrlAliasHandler.php [hunk 2]
```diff
@@ -288,8 +288,28 @@ public function locationDeleted($locationId)
...
+    {
+        $locationCache = $this->cache->getItem('urlAlias', 'location', $locationId);
+
+        if ($locationCache->isMiss()) {
+            // we need to clear all if we don't have location id in cache
+            $this->cache->clear('urlAlias');
+        } else {
+            $urlAliasesId = $locationCache->get();
+
+            foreach ((array)$urlAliasesId as $eachUrlAliasId) {
```
Comment (andrerom) [inline eZ/Publish/Core/Persistence/Cache/UrlAliasHandler.php]: `$eachUrlAliasId` is not that descriptive of what it is, lets do the similar as existing code:

```
$urlAliasIds = $locationCache->get();

foreach ($urlAliasIds as $urlAliasId) {
   $this->cache->clear('urlAlias', $urlAliasId);
```

`$urlAliasIds` could also be `$urlAliasIdList` or similar.

Comment (galileo) [inline eZ/Publish/Core/Persistence/Cache/UrlAliasHandler.php]: The cast is here, because I don't know if the value is stored always in array so to ensure this foreach method works as expected I casted it, I will remove it.

File: eZ/Publish/Core/Persistence/Cache/UrlAliasHandler.php (hunk 1 above)
Comment (galileo) [inline eZ/Publish/Core/Persistence/Cache/UrlAliasHandler.php]: It just an residue from cache warmup, i will change this. Also I created new jira ticket for this cache warmup feature.

Url alias cache warmup: https://jira.ez.no/browse/EZP-25773

File: eZ/Publish/Core/Persistence/Cache/UrlAliasHandler.php (hunk 2 above)
Comment (galileo) [inline eZ/Publish/Core/Persistence/Cache/UrlAliasHandler.php]: @andrerom 

So as I excepted, in unit test there are places where the get method returns for example null or 0. And this can not be used in foreach, so if I should remove the array cast, what is the best next step? 

1. Change this test and mock it to return array?
2. Add condition for run only if the given value is an array?
3. Cast the value to array?

**Comment by galileo** [comment]
I changed the code mentioned by you, but left for now the array cast, and now I wait for the aprove of  #1649  because I need this to end this PR.

ping @andrerom

File: eZ/Publish/Core/Persistence/Cache/UrlAliasHandler.php (hunk 2 above)
Comment (andrerom) [inline eZ/Publish/Core/Persistence/Cache/UrlAliasHandler.php]: Whne you say tests I guess unit tests? If so it's just missing impl in tests, and you should adjust the test.

Comment (galileo) [inline eZ/Publish/Core/Persistence/Cache/UrlAliasHandler.php]: Ohhh ok, now all is clear I will change this implementation. Thank you for explanation.

**Comment by andrerom** [comment]
+1

**Comment by bdunogier** [comment]
+1, here we go.
